### ISSUE TITLE
Inefficient, newly introduced pause in train_on_batch()

### ISSUE THREAD

**Issue by wulabs** [fchollet/keras#6243]
Since upgrading to Keras 2 and latest Theano (last updated ~Jan '17), the train_on_batch() call now takes a ~10 second pause after approximately the first 10 batches for Convolutional layers.  Is there a reason for this pause, and is there a way to get around it?

**Comment by fchollet** [comment]
Try using a profiler to determine which specific lines of code are are involved in a potential "pause". Please report back with a line-by-line breakdown of time taken and a code example for reproduction.

**Comment by wulabs** [comment]
The offending call is _make_train_function in training.py.

As seen from profile output below, this is only affected by Theano. I am using Theano master branch, commit 6c3d8b4302eab84c7ad8ad375c0d4061f675a11e from April 13th.

Below is profile output of a single-layer 64 neuron dense layer for 1 mini-batch on Cuda device. Blacked out areas for proprietary info.  First screenshot is running with Theano backend. Second screenshot is running with Tensorflow backend. Both have the same highlighted call.

I will work on reproducible code. Currently this is embedded in a large proprietary project.

![theano](https://cloud.githubusercontent.com/assets/596484/25103606/1487255c-2372-11e7-9902-92843d08c5ff.png)
![tensorflow](https://cloud.githubusercontent.com/assets/596484/25103607/1490b14e-2372-11e7-81bd-b9c3231a3e58.png)

**Comment by fchollet** [comment]
Normally, `_make_train_function` is only called once, then the resulting
function is cached and reused. Why would it be called several times in your
case?

**Comment by wulabs** [comment]
It seems `_make_train_function` is called for every call of `train_on_batch`. We are using this in a RL DQN so there are usually multiple calls to `train_on_batch`, one for each step in an epoch. The model is compiled only once, but inputs and outputs change for each epoch. For this use case, should I still only see one call to `_make_train_function`?

In the previous profiler generation there was an off-by-1 bug causing 2 epoch steps.  Attached is updated profile output for only 1 iteration, and hence one call to `train_on_batch`.  The `_make_train_function` delay still exists, but only for the first call (as can be seen between the difference in these and last screenshots). First screenshot is Theano-Cuda. Second is Tensorflow-Cuda.

![theano_cuda](https://cloud.githubusercontent.com/assets/596484/25111633/01ecab36-23a1-11e7-8fb2-1f6be00d0b13.png)
![tensorflow_cuda](https://cloud.githubusercontent.com/assets/596484/25111637/03d9da40-23a1-11e7-8c8d-ac6baf7cf0c5.png)

**Comment by fchollet** [comment]
What's your code like? Do you have a snippet to reproduce?

**Comment by wulabs** [comment]
Reproducible code below. Attached profile output screenshots for this snippet.

`"""
Testing for Theano.Cuda _make_train_function delay bug.
"""
import numpy as np
import random
np.random.seed(7)
random.seed(7)

import keras
import keras.layers as k_layers
from keras.layers import  Dense, Input
from keras.models import Model as Keras_Model

x_train = np.array([[ 0.4995946,  0.4995946,  0.4995946,  0.4451019,  0.4451019, 0.4451019]])
targets = np.array([[ 0.00432717,  0. ]])
x_input = Input(shape=(x_train.shape[1],))

x = Dense(64, kernel_initializer="uniform")(x_input)

output = Dense(2, activation='linear', name='output_layer')(x)

model = Keras_Model(inputs=x_input, outputs=output)

model.compile(
    optimizer="adam",
    loss="mse",
    metrics=["accuracy"])

metrics = model.train_on_batch(x_train, targets)
print metrics`


![theano_cuda_snippet](https://cloud.githubusercontent.com/assets/596484/25112799/6e598d0a-23a8-11e7-9488-acb7a72cceb4.png)

![tensorflow_cuda_snippet](https://cloud.githubusercontent.com/assets/596484/25112798/6e5755f8-23a8-11e7-966e-5bc2305b1c4c.png)
